### ISSUE TITLE
Overspent ex budget in some minting scripts

### ISSUE THREAD

**Issue by christianschmitz** [Hyperion-BT/helios#57]
Minting script with the following IR throws an Overspent Budget error when submitted through blockfrost:

```
01  (redeemer, ctx) -> {
02    __core__ifThenElse(
03      __core__ifThenElse(
04        __core__equalsInteger(__core__fstPair(__core__unConstrData(redeemer)), 0),
05        () -> {
06          (_) -> {
07            __core__equalsData(ctx, ctx)
08          }
09        },
10        () -> {
11          (_) -> {
12            true
13          }
14        }
15      )()(redeemer),
16      () -> {
17        ()
18      },
19      () -> {
20        error()
21      }
22    )()
23  }
```

**Comment by christianschmitz** [comment]
(narrowing it down) the following script also throws the error:
```
minting factory

enum Redeemer {
    Create
    Delete
}

func main(redeemer: Redeemer, context: ScriptContext) -> Bool {
    redeemer == redeemer && context == context
}
```

With IR:
```
01  (__helios__common____eq) -> {
02    (redeemer, ctx) -> {
03      __core__ifThenElse(
04        __core__ifThenElse(
05          __helios__common____eq(redeemer)(redeemer),
06          () -> {
07            __helios__common____eq(ctx)(ctx)
08          },
09          () -> {
10            false
11          }
12        )(),
13        () -> {
14          ()
15        },
16        () -> {
17          error()
18        }
19      )()
20    }
21  }(/*__helios__common____eq*/(self) -> {
22    (other) -> {
23      __core__equalsData(self, other)
24    }
25  })
```

**Comment by christianschmitz** [comment]
(narrowing it down further) the following script also throws the error:
```
minting factory

struct Redeemer {
  i: Int
}

func main(redeemer: Redeemer, context: ScriptContext) -> Bool {
    redeemer == redeemer && context == context
}
```

(identical IR as above)

**Comment by christianschmitz** [comment]
The following does NOT throw the error:
```
minting factory

struct Redeemer {
  i: Int
}

func main(redeemer: Redeemer) -> Bool {
    redeemer == redeemer && redeemer == redeemer
}
```

**Comment by christianschmitz** [comment]
The following throws the error:

```
minting factory

struct Redeemer {
  b: Bool
}

func main(redeemer: Redeemer, context: ScriptContext) -> Bool {
    redeemer.b && context == context
}
```

With IR:
```
01  (redeemer, ctx) -> {
02    __core__ifThenElse(
03      __core__ifThenElse(
04        __core__ifThenElse(
05          __core__equalsInteger(__core__fstPair(__core__unConstrData(redeemer)), 0),
06          false,
07          true
08        ),
09        () -> {
10          __core__equalsData(ctx, ctx)
11        },
12        () -> {
13          false
14        }
15      )(),
16      () -> {
17        ()
18      },
19      () -> {
20        error()
21      }
22    )()
23  }
```

**Comment by christianschmitz** [comment]
Following does NOT throw an error:

```
minting factory

struct Redeemer {
  b: Bool
}

func main(redeemer: Redeemer) -> Bool {
    redeemer.b && redeemer == redeemer
}
```

With UPLC:
```
(program 1.0.0 
  (Λ redeemer -> 
    (Λ _ -> 
      (force [
        [
          [
            (force (builtin ifThenElse))
            (force [
                [
                [(force (builtin ifThenElse))
                [[[(force (builtin ifThenElse))
                [[(builtin equalsInteger)
                [(force (force (builtin fstPair)))
                [(builtin unConstrData) x2]]] 0]] false] true]
                ]
                (delay [[(builtin equalsData) x2] x2])
              ]
              (delay false)])
            ] (delay ())
          ] 
          (delay (error))
        ]
      )
    )
  )
)
```

**Comment by christianschmitz** [comment]
(narrowing it down further) the following throws the error:
```
minting factory

func main(ctx: ScriptContext) -> Bool {
    ctx == ctx
}
```

With IR:
```
01  (_, ctx) -> {
02    __core__ifThenElse(
03      __core__equalsData(ctx, ctx),
04      () -> {
05        ()
06      },
07      () -> {
08        error()
09      }
10    )()
11  }
```

With UPLC:
```
(program 1.0.0 (Λ _ -> (Λ ctx -> (force [[[(force (builtin ifThenElse)) [[(builtin equalsData) x1] x1]] (delay ())] (delay (error))]))))
```

**Comment by christianschmitz** [comment]
I've been able to pin it down to a ScriptContext that differs from ex budget calculation to actual tx submission stage (due to final balancing happening after ex budget calculation).

Solution is to add a dummy input and dummy output to the ScriptContext used for ex budget calculation.

Fix has been add to v0.12.7 of Helios
